sktime/classification/distance_based/_shape_dtw.py

```diff
@@ -0,0 +1,477 @@
+            sdf = self.shape_descriptor_function
+            sdfs = self.shape_descriptor_functions
+            if sdfs is None or not (len(sdfs) == 2):
+                raise ValueError("When using 'compound', " +
+                                 "shape_descriptor_functions must be a " +
+                                 "string array of length 2.")
+            mp = self.metric_params
+
+            grid = GridSearchCV(
+                estimator=ShapeDTW(n_neighbors=n,
+                                   subsequence_length=sl,
+                                   shape_descriptor_function=sdf,
+                                   shape_descriptor_functions=sdfs,
+                                   metric_params=mp),
+                param_grid=self._param_matrix,
+                cv=KFold(n_splits=10, shuffle=True),
+                scoring='accuracy'
+            )
+            grid.fit(X, y)
+            self.weighting_factor = \
+                grid.best_params_['metric_params']['weighting_factor']
+
+    def predict_proba(self, X):
+        """
+        Function to perform predictions on the testing data X. This function
+        returns the probabilities for each class.
+         
+        Parameters
+        ----------
+        X - pandas dataframe of testing data of shape [num_insts,1].
+
+        Returns
+        -------
+        output : numpy array of shape =
+                [num_insts, num_classes] of probabilities
+        """
+        X = check_X(X, enforce_univariate=False)
+
+        self.testData = X
+
+        # get the number of attributes and instances
+        num_atts = self.testData.shape[1]
+        num_insts = self.testData.shape[0]
+
+        # Convert testing data into a list of subsequences
+        st = SlidingWindowSegmenter(self.subsequence_length)
+        st.fit(self.testData)
+        self.sequences = st.transform(self.testData)
+
+        self.testData = self.sequences
+
+        # Create the testing data by finding the shape descriptors
+        self.testData = self._generate_shape_descriptors(
+                            self.sequences, num_insts, num_atts)
+
+        # Classify the test data
+        return self.knn.predict_proba(self.testData)
+
+    def predict(self, X):
+        """
+        Find predictions for all cases in X.
+        Could do a wrap function for predict_proba,
+        but this will do for now.
+        ----------
+        X : The testing input samples of shape [num_insts,1].
+
+        Returns
+        -------
+        output : numpy array of shape = [num_insts]
+        """
+        X = check_X(X, enforce_univariate=False)
+        self.testData = X
+
+        # get the number of attributes and instances
+        num_atts = self.testData.shape[1]
+        num_insts = self.testData.shape[0]
+
+        # Convert testing data into a list of subsequences
+        st = SlidingWindowSegmenter(self.subsequence_length)
+        st.fit(self.testData)
```

Comment:
I find it strange to see this being fitted in predict, could you leave a short comment why we fit this in predict rather than using a fitted instance from fit?

Reply:
Yeah so from your previous comment I've made a _preprocess method and that's what that is doing, it should make more sense now with the _preprocess method because ShapeDTW has to transform the test data as well as the training data

Reply:
The SlidingWindowSegmenter is a deterministic transform, in the sense that noting is being estimated or fitted to the data, and so fit is empty, right? But technically, shouldn't you call fit and transform in fit and then only call transform on the fitted instance in predict? It's like a pipeline no?

Reply:
Yeah exactly yeah fit is baiscally empty. oh okay I think I see what you mean now yeah I was a bit worried that just calling transform in predict might break it but I can make that change so that it only transforms in predict and remove the fit call